src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1587 @@
+
+use bitcoin_hashes::{Hash, hash160, sha256, sha256d, ripemd160};
+use miniscript::astelem::AstElem;
+use ::{ToPublicKeyHash};
+use secp256k1::{self, Signature, VerifyOnly};
+use ::{ToPublicKey, Descriptor};
+use ::{bitcoin};
+use error;
+use fmt;
+/// Detailed Error type for Interpreter
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Error {
+    /// Unexpected Stack End, caused by popping extra elements from stack
+    UnexpectedStackEnd,
+    /// Unexpected Stack Push `StackElement::Push` element when the interpreter
+    /// was expecting a stack boolean `StackElement::Satisfied` or
+    /// `StackElement::Dissatisfied`
+    UnexpectedStackElementPush,
+    /// Verify expects stack top element exactly to be `StackElement::Satisfied`.
+    /// This error is raised even if the stack top is `StackElement::Push`.
+    VerifyFailed,
+    /// MultiSig missing at least `1` witness elements out of `k + 1` required
+    InsufficientSignaturesMultiSig,
+    /// MultiSig requires 1 extra zero element apart from the `k` signatures
+    MissingExtraZeroMultiSig,
+    /// Script abortion because of incorrect dissatisfaction for multisig.
+    /// Any input witness apart from sat(0 sig ...) or nsat(0 0 ..) leads to
+    /// this error. This is network standardness assumption and miniscript only
+    /// supports standard scripts
+    MultiSigEvaluationError,
+    /// General Interpreter error.
+    CouldNotEvaluate,
+    /// Script abortion because of incorrect dissatisfaction for Checksig.
+    /// Any input witness apart from sat(sig) or nsat(0) leads to
+    /// this error. This is network standardness assumption and miniscript only
+    /// supports standard scripts
+    PkEvaluationError(bitcoin::PublicKey),
+    /// Miniscript requires the entire top level script to be satisfied.
+    ScriptSatisfactionError,
+    /// The Public Key hash check for the given pubkey. This occurs in `PkH`
+    /// node when the given key does not match to Hash in script.
+    PkHashVerifyFail(hash160::Hash),
+    /// Parse Error while parsing a `StackElement::Push` as a Pubkey. Both
+    /// 33 byte and 65 bytes are supported.
+    PubkeyParseEror,
+    /// The preimage to the hash function must be exactly 32 bytes.
+    HashPreimageLenghtMismatch,
+    /// Got `StackElement::Satisfied` or `StackElement::DisSatisfied` when the
+    /// interpreter was expecting `StackElement::Push`
+    UnexpectedStackBoolean,
+    /// Could not satisfy, relative locktime not met
+    RelativeLocktimeNotMet(u32),
+    /// Could not satisfy, absolute locktime not met
+    AbsoluteLocktimeNotMet(u32),
+    /// Forward-secp related errors
+    Secp(secp256k1::Error),
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        ""
+    }
+    fn cause(&self) -> Option<&error::Error> {
+        None
```

Comment:
Should return the secp error

Reply:
On other cases, None? I am not really sure how to do this.